### ISSUE TITLE
"debug" dependency breaking stuff

### ISSUE THREAD

**Issue by Kronuz** [expressjs/body-parser#261]
Dependency locked to a version (`"debug": "2.6.7"`) is breaking our code because it makes all other `"debug": "^2.6.8"` use v2.6.7 instead (which has a problem). The dependency was updated to be "2.6.8" in master, but not in npm. Still, however, this delays the problem to next version (2.6.9 perhaps) Why not better use `"debug": "^2.6.8"` instead? (which would fix the issue)

**Comment by dougwilson** [comment]
Hi @Kronuz you are encountering an npm issue; npm should allow this module to keep it's own version of debug and not cause an invalid version in your other modules. We used to allow a range, but `debug` has periodically released bugs affecting all our users with no recourse. We my policy only exact pin any module we don't control the releases of.

**Comment by dougwilson** [comment]
For example, here is how a functional version of `npm` will install what you described. Given the following `package.json`:
```
$ cat package.json 
{
  "name": "body-parser-261",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "debug": "^2.6.8",
    "body-parser": "1.17.2"
  },
  "author": "",
  "license": "ISC"
}
```

Here is the `npm install` output and tree:
```
$ npm install
npm WARN body-parser-261@1.0.0 No description
npm WARN body-parser-261@1.0.0 No repository field.

added 21 packages in 2.224s

$ npm ls
body-parser-261@1.0.0 /body-parser-261
├─┬ body-parser@1.17.2
│ ├── bytes@2.4.0
│ ├── content-type@1.0.2
│ ├─┬ debug@2.6.7
│ │ └── ms@2.0.0 deduped
│ ├── depd@1.1.1
│ ├─┬ http-errors@1.6.2
│ │ ├── depd@1.1.1 deduped
│ │ ├── inherits@2.0.3
│ │ ├── setprototypeof@1.0.3
│ │ └── statuses@1.3.1
│ ├── iconv-lite@0.4.15
│ ├─┬ on-finished@2.3.0
│ │ └── ee-first@1.1.1
│ ├── qs@6.4.0
│ ├─┬ raw-body@2.2.0
│ │ ├── bytes@2.4.0 deduped
│ │ ├── iconv-lite@0.4.15 deduped
│ │ └── unpipe@1.0.0
│ └─┬ type-is@1.6.15
│   ├── media-typer@0.3.0
│   └─┬ mime-types@2.1.17
│     └── mime-db@1.30.0
└─┬ debug@2.6.8
  └── ms@2.0.0
```

You can see that `debug` 2.6.8 is installed at the top level, to honor the `"debug": "^2.6.8"` and a second `debug` is installed within `body-parser`. You may need to report the bug to npm or upgrade your npm version to a non-buggy version.

**Comment by Kronuz** [comment]
I understand and agree it's an issue with npm. Other packages listing "debug": "^2.6.8" should use 2.6.8 leaving that body-parser with 2.6.7. For the moment, it'd be great and I'd appreciate if you could release a new version of body-parser with the 2.6.8 which fixes the issues I was having (even if pinned)

**Comment by dougwilson** [comment]
Yea, I can make a release when I get back from va-cay in a couple days, no problem 👍

**Comment by dougwilson** [comment]
Any objection to publishing with a requirement of 3.0.1?

**Comment by Kronuz** [comment]
Not sure, everyone seem to still be using 2.6.8

**Comment by dougwilson** [comment]
I meant for your isdue. I've been upgrading to 3.0.1 on everything so far without issue. Just want to make sure that it doesn't just make your problem still be there. I can also verify if you can provide the steps to replicate your issue.

**Comment by Kronuz** [comment]
@dougwilson, it'd be better if it's in 2.6.8 for me (it's less disruptive)

**Comment by dougwilson** [comment]
Thanks for your response! I will publish tonight. Do you think I can ever bump this module to debug 3.x?

**Comment by dougwilson** [comment]
I would rather not release a new version with an outdated debug dependency, especially after going through the two times of upgrading debug everywhere because it had a security issue.

**Comment by Kronuz** [comment]
It probably would be fine, but the thing is debug crosses major version... so I'll end up with two versions of debug everywhere ('cause all the other packages haven't upgraded yet) ...maybe release with 2.6.8 and wait a bit until others (morgan, mocha, send, express, eslint, compression, etc) start using 3.0.1?

**Comment by dougwilson** [comment]
Hi @Kronuz but unless they all release a new version at exactly the same time, then there will always been some with one version and some with the other. Unless there is a good reason, or you can provide a plan of action on how to actually get this module onto 3.0.1, I'm going to just release new version with 3.0.1 to get the ball rolling here. For example, if each of those modules waits for all the others to update, it would be a stalemate... no one can upgrade.

**Comment by Kronuz** [comment]
That is true. I guess 3.0.1 would be fine.

**Comment by Kronuz** [comment]
@dougwilson, didn't get the change for a new release or is it something that blocked it?

**Comment by dougwilson** [comment]
I'm working on it, just had an interruption to investigate a memory leak in Express that took up my time yesterday I would have spent on it. Then over night a repost of a potential security vulnerability which took more time.

**Comment by dougwilson** [comment]
So looking into the debug version, I decided that it was best to stick with the 2.x for now. I published 1.18.0 with the new version.

**Comment by Kronuz** [comment]
Great! Thank you so much! 😁👍
